cli/client/dcoscli.go

```diff
@@ -49,10 +58,112 @@ func RequiredCLIConfigValue(name string, description string, errorInstruction st
+// YAML config file directly before falling back to querying disk
+func cliConfigValue(name string) (string, error) {
+	// dcos-cli allows providing a custom envvar for configs, we honor that here:
+	configDir := os.Getenv("DCOS_DIR")
+	if len(configDir) == 0 {
+		homeDir := os.Getenv("HOME")
+		if len(homeDir) != 0 {
+			configDir = path.Join(homeDir, ".dcos")
+		}
+	}
+
+	if len(configDir) != 0 {
+		diskConfig, err := cliDiskConfig(configDir)
+		if err == nil {
+			return cliDiskConfigValue(diskConfig, name)
+		} else if config.Verbose {
+			PrintMessage("No cluster config found, falling back to querying CLI: %s", err.Error())
+		}
+	} else if config.Verbose {
+		PrintMessage("Unable to resolve CLI config directory (DCOS_DIR or HOME/.dcos), falling back to running CLI command.")
+	}
+
+	// Fall back to querying the CLI directly: slower but works on older CLIs.
+	return RunCLICommand("config", "show", name)
+}
+
+// cliDiskConfigValue returns a value from the user's CLI configuration on disk,
+// or an error if the value could not be retrieved.
+func cliDiskConfigValue(diskConfig map[string]interface{}, name string) (string, error) {
+	// map "foo.bar.baz" in name to '[foo]\nbar.baz = "..."' in config file
+	tokens := strings.SplitN(name, ".", 2)
+	if len(tokens) != 2 {
+		return "", fmt.Errorf("Unable to tokenize config value into 'section.name': %s => %s", name, tokens)
+	}
+
+	rawSection, ok := diskConfig[tokens[0]]
+	if !ok {
+		return "", fmt.Errorf("No section named '%s' when fetching '%s'", tokens[0], name)
+	}
+	section, ok := rawSection.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("Unable to convert content of section '%s' into map", tokens[0])
+	}
+	rawValue, ok := section[tokens[1]]
+	if !ok {
+		return "", fmt.Errorf("Unable to retrieve value '%s' from section '%s'", tokens[1], tokens[0])
+	}
+	value, ok := rawValue.(string)
+	if !ok {
+		return "", fmt.Errorf("Unable to convert value '%s' from section '%s' into string", tokens[1], tokens[0])
+	}
+	return value, nil
+}
+
+// cliDiskConfig returns the content of the user's CLI configuration on disk,
+// or an error if the configuration could not be retrieved.
+func cliDiskConfig(configDir string) (map[string]interface{}, error) {
+	if cachedConfig != nil {
+		return cachedConfig, nil
+	}
+
+	// Find the "dcos.toml" config file:
+	configPath := ""
+
+	// Look for new-style multicluster configs:
+	clustersPath := path.Join(configDir, "clusters")
+	clusters, err := ioutil.ReadDir(clustersPath)
+	if err == nil {
+		for _, cluster := range clusters {
+			if !cluster.IsDir() {
+				continue // not a directory
+			}
+			if _, err := os.Stat(path.Join(clustersPath, cluster.Name(), "attached")); os.IsNotExist(err) {
+				continue // not the attached/active cluster
+			}
+			configPath = path.Join(clustersPath, cluster.Name(), "dcos.toml")
+			break
+		}
+	}
+
+	// Fall back to old-style single config:
+	if configPath == "" {
+		configPath = path.Join(configDir, "dcos.toml")
```

Comment:
I can't tell, is this exercised in any of the tests?

Reply:
Added a test case for a directory which only has an old config